hwtracer/src/decode/ykpt/mod.rs

```diff
@@ -19,29 +38,232 @@ impl TraceDecoder for YkPTTraceDecoder {
+                    if self.tnts.is_empty() {
+                        if self.seek_tnt()? == false {
+                            // The packet stream was exhausted.
+                            return Ok(None);
+                        }
+                    }
+                    // Find where to go next based on whether the trace took this branch or not.
+                    let target_off = if self.tnts.pop_front().unwrap() {
+                        *taken_target
+                    } else {
+                        if let Some(ntt) = not_taken_target {
+                            *ntt
+                        } else {
+                            // Divergent control flow.
+                            panic!();
+                        }
+                    };
+                    self.cur_loc = ObjLoc::MainObj(target_off);
+                    return Ok(Some(self.lookup_block_from_main_bin_offset(target_off)));
+                }
+                SuccessorKind::Dynamic => {
+                    // We can only know the successor via a TIP update in a packet. Fall through to
+                    // `self.seek_tip()`.
+                    self.cur_loc = ObjLoc::OtherObjOrUnknown;
+                }
+            }
+        }
+
+        // If we get here then we can't statically figure out where to go next. A packet is going
+        // to have to tell us what to do.
+        self.seek_tip()
+    }
+
+    /// Keep decoding packets until we have some TNT decisions buffered.
+    fn seek_tnt(&mut self) -> Result<bool, HWTracerError> {
+        while self.tnts.is_empty() {
+            let pkt = self.packet()?; // Potentially populates `self.tnts`.
+            if pkt.is_none() {
+                return Ok(false); // Packet stream exhausted.
+            }
+        }
+        Ok(true)
+    }
+
+    /// Keep decoding packets until we receive a TIP update.
+    fn seek_tip(&mut self) -> Result<Option<Block>, HWTracerError> {
+        debug_assert_eq!(self.cur_loc, ObjLoc::OtherObjOrUnknown);
+        loop {
+            if let Some(_) = self.packet()? {
+                if let ObjLoc::MainObj(off) = self.cur_loc {
+                    return Ok(Some(self.lookup_block_from_main_bin_offset(off)));
+                }
+            } else {
+                return Ok(None); // No more packets.
+            }
+        }
+    }
+
+    /// Fetch the next packet and update iterator state.
+    fn packet(&mut self) -> Result<Option<Packet>, HWTracerError> {
+        let ret = if let Some(pkt_or_err) = self.parser.next() {
+            let pkt = pkt_or_err?;
+            if pkt.kind() == PacketKind::TIPPGE {
+                debug_assert!(!self.pge);
+                self.pge = true;
+            }
+            if pkt.kind() == PacketKind::TIPPGD {
+                debug_assert!(self.pge);
+                self.pge = false;
+            }
+            if let Some(vaddr) = pkt.target_ip() {
```

Comment:
Can this be true *and* the packet be TIPPGE or TIPPGD?

Reply:
Yes. Both of those packets deliver a TIP update.

![image](https://user-images.githubusercontent.com/604955/203814533-0b2b694d-c4f9-46ec-abda-2d8b729987ce.png)

![image](https://user-images.githubusercontent.com/604955/203814630-9a38b26e-74ca-409e-9cf4-b60ccd0ebb29.png)

Reply:
Got it!